### PR TITLE
TINY-9275: Custom tooltips

### PR DESCRIPTION
**Pull request by shanmen-tiny**
Related Ticket:  TINY-9275

Description of Changes:
* New behaviour: `Tooltipping` (the spike PR #8114) 
* Add custom tooltip on various UI: toolbar buttons, dropdowns (no label), icon buttons, color pickers, split buttons and choice buttons
* Tooltip is rendered immediately on focus using keyboard navigation, there will be a slight delay when using a mouse to hover
* Good to also review the changes that I've recently made, mainly for reverting the changes that the title attribute was removed, I've added a comment and linked a ticket for places where we need to remove the title attribute.
* https://github.com/tinymce/tinymce/compare/5967e8e458f7783af1e7150255eb4fb3fb2f84c5...3e0198b0c39fe11753b6056be720a3d2ecdb0b6e 	
* The changelog will be updated again, when the title attribute is removed and tests are updated in an upcoming ticket https://ephocks.atlassian.net/browse/TINY-10453
* Will also look to lower the `tooltipDelay` in an upcoming [TINY-10475](https://ephocks.atlassian.net/browse/TINY-10475)

Pre-checks:
* [x] Changelog entry added
* [x] Tests have been added (if applicable)
* [x] Branch prefixed with `feature/`, `hotfix/` or `spike/`

Review:
* [x] Milestone set
* [x] Docs ticket created (if applicable)

GitHub issues (if applicable):

[TINY-10475]: https://ephocks.atlassian.net/browse/TINY-10475?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/item/build/AutocompleteMenuItem.ts [hunk 1]
```diff
@@ -89,6 +89,7 @@ const renderAutocompleteItem = (
     onAction: (_api) => onItemValueHandler(spec.value, spec.meta),
     onSetup: Fun.constant(Fun.noop),
     triggersSubmenu: false,
+    // TINY-9125: Hmm. This is going to be a problem.
```
Comment (TheSpyder) [inline modules/tinymce/src/themes/silver/main/ts/ui/menus/item/build/AutocompleteMenuItem.ts:92]: ?

File: modules/tinymce/src/themes/silver/main/ts/backstage/TooltipsBackstage.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  const getConfig = (spec: { tooltipText: string }) => ({
+    delayForShow: () => alreadyShowingTooltips() ? intervalDelay : tooltipDelay,
+    delayForHide: Fun.constant(tooltipDelay),
+    exclusive: true,
+    lazySink: getSink,
+    /*
+      <div class="tox-tooltip tox-tooltip--down">
+  <div class="tox-tooltip__body">Are you hanging on the edge of your seat? </div>
+  <i class="tox-tooltip__arrow"></i>
+  </div>*/
```
Comment (TheSpyder) [inline modules/tinymce/src/themes/silver/main/ts/backstage/TooltipsBackstage.ts:28]: can probably remove this?

**Review by TheSpyder — APPROVED** [review]

File: modules/tinymce/src/themes/silver/test/ts/browser/editor/TooltipTest.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+import { describe, it } from '@ephox/bedrock-client';
+import { SugarBody, SugarElement } from '@ephox/sugar';
+import { TinyDom, TinyHooks, TinyUiActions } from '@ephox/wrap-mcagar';
+import { assert } from 'chai';
+
+import Editor from 'tinymce/core/api/Editor';
+import ImagePlugin from 'tinymce/plugins/image/Plugin';
+import LinkPlugin from 'tinymce/plugins/link/Plugin';
+
+describe('browser.tinymce.themes.silver.editor.Tooltip', () => {
```
Comment (ltrouton) [inline modules/tinymce/src/themes/silver/test/ts/browser/editor/TooltipTest.ts:11]:
```suggestion
describe('browser.tinymce.themes.silver.editor.TooltipTest', () => {
```

File: modules/tinymce/src/themes/silver/main/ts/ui/general/Button.ts [hunk 4]
```diff
@@ -138,12 +141,26 @@ export const renderButtonSpec = (
     tag: 'button',
     classes,
     attributes: {
-      title: translatedText // TODO: tooltips AP-213
+      title: translatedText
```
Comment (ltrouton) [inline modules/tinymce/src/themes/silver/main/ts/ui/general/Button.ts:144]: I assume we want a TODO to remove this later?

**Review by ltrouton — COMMENTED** [review]
Some initial comments

File: modules/oxide/src/less/skins/ui/dark/skin.less [hunk 5]
```diff
@@ -6,6 +6,7 @@
 
 @background-color: #222F3E;
 @tinymce-separator-color: rgba(255,255,255,.15);
+@tooltip-background-color: #324053;
```
[the remark this answers is not in the document]
Comment (shanmen-tiny) [reply modules/oxide/src/less/skins/ui/dark/skin.less:9]: I have created a ticket here, [TINY-10484](https://ephocks.atlassian.net/browse/TINY-10484)

File: modules/oxide/src/less/theme/components/tooltip/tooltip.less [hunk 6]
```diff
@@ -3,22 +3,31 @@
 //
 
 @tooltip-arrow-size: @base-value / 2;
-@tooltip-background-color: contrast(@background-color, @color-black, lighten(@background-color, 15%));
-@tooltip-border-radius: @panel-border-radius;
-@tooltip-box-shadow: 0 2px 4px fade(@color-black, 30%);
-@tooltip-font-size: @font-size-sm;
-@tooltip-font-weight: @font-weight-normal;
+@tooltip-background-color: @color-black;
```
[the remark this answers is not in the document]
Comment (shanmen-tiny) [reply modules/oxide/src/less/theme/components/tooltip/tooltip.less:6]: Crossed check the figma and the variable, both are: #222F3E

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/item/build/AutocompleteMenuItem.ts (hunk 1 above)
Comment (shanmen-tiny) [reply modules/tinymce/src/themes/silver/main/ts/ui/menus/item/build/AutocompleteMenuItem.ts]: Change the comment to fix it in TINY-9638

File: modules/alloy/docs/behaviours/Tooltipping.adoc [hunk 7]
```diff
@@ -0,0 +1,82 @@
+== General Overview
+
+The `Tooltipping` behaviour allows:
+
+* the entire mothership to show just one tooltip at a time through `exclusive`. This works by broadcasting a message that all Tooltipping behaviours listen to.
+* have two different modes of tooltipping:
+  * normal: does stuff
+  * follow-highlight: does other stuff.
```
[the remark this answers is not in the document]
Comment (shanmen-tiny) [reply modules/alloy/docs/behaviours/Tooltipping.adoc]: Updated docs

File: modules/tinymce/src/themes/silver/main/ts/ui/general/Button.ts (hunk 4 above)
Comment (shanmen-tiny) [reply modules/tinymce/src/themes/silver/main/ts/ui/general/Button.ts:145]: Added

**Review by ltrouton — APPROVED** [review]

**Pull request merged**